external/fv3fit/fv3fit/reservoir/transformers/transformer.py

```diff
@@ -32,7 +38,250 @@ def decode(self, x: ArrayLike) -> Sequence[ArrayLike]:
+class _ScaleSpatialConcatZTransformer(Transformer):
+    """
+    Limited utility transformer that specifically takes in a
+    sequence of data with the x,y,z trailing features, normalizes
+    the data per individual xyz feature, and then concatenates along the
+    z dimension.
+
+    Requires that all input arrays have the same x,y,z features. Use
+    the build function below to construct this transformer.
+
+    Attributes:
+        center: mean of each variable, shape (x * y * z)
+        scale: standard deviation of each variable, shape (x * y * z)
+        spatial_features: x,y,z dimensions of the input data
+        num_variables: number of variables in the input data
+        mask: optional mask to apply to the encoded output, shape
+            (x, y, z * num_variables)
+    """
+
+    _CONFIG_NAME = "scale_spatial_concat_z_transformer.yaml"
+    _SCALE_NDARRAY = "scale.npy"
+    _CENTER_NDARRAY = "center.npy"
+    _MASK_NDARRAY = "mask.npy"
+    _EPSILON = 1.0e-7
+
+    def __init__(
+        self,
+        center: np.ndarray,
+        scale: np.ndarray,
+        spatial_features: Sequence[int],
+        num_variables: int,
+        mask: Optional[np.ndarray] = None,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self._spatial_features = spatial_features
+        self._num_variables = num_variables
+        self._norm_layer = NormLayer(center=center, scale=scale, epsilon=self._EPSILON)
+        self._mask = mask
+
+    @property
+    def n_latent_dims(self):
+        return self._num_variables * self._spatial_features[-1]
+
+    @property
+    def _flat_spatial_len(self):
+        return np.product(self._spatial_features)
+
+    @property
+    def _stacked_flat_spatial_split_idxs(self):
+        return [self._flat_spatial_len * i for i in range(1, self._num_variables)]
+
+    @property
+    def _z_dim_split_idxs(self):
+        return [self._spatial_features[-1] * i for i in range(1, self._num_variables)]
+
+    def _check_consistent_xyz(self, input_arrs: Sequence[np.ndarray]):
+        if len(input_arrs) != self._num_variables:
+            raise ValueError(
+                f"Expected {self._num_variables} input arrays but got {len(input_arrs)}"
+            )
+
+        for i, arr in enumerate(input_arrs):
+            if arr.shape[-3:] != self._spatial_features:
+                raise ValueError(
+                    "All arrays must have the same x,y,z features. "
+                    f"Expected {self._spatial_features} but got {arr.shape[-3:]} "
+                    f"for array {i}."
+                )
+
+    def encode_txyz(self, input_arrs: Sequence[ndarray]) -> ndarray:
+        # Overloads the original transformer encode_txyz since we want to
+        # handle the xyz dimensions differently.
+
+        self._check_consistent_xyz(input_arrs)
+
+        leading_dims = input_arrs[0].shape[:-3]
+        # stack xyz
+        spatial_last_dim = [tf.reshape(arr, (*leading_dims, -1)) for arr in input_arrs]
```

Comment:
Since this is already a fairly singularly focused class, I'm hesitant to write out too many subunits for small tasks that wont be generalized.  I agree about the non-uniform arrays not being informative of correctness.  Instead, to verify that everything is encoded/decoded properly, I adjusted the sample arrays to use `np.arange`.  Each xyz element will be unique in this case, so we can't pass tests without correct reshaping.

Reply:
See d2d64696f

Reply:
See [d2d6469](https://github.com/ai2cm/fv3net/pull/2382/commits/d2d64696f520d2d95dd34f78483735c2bb60a354)

Reply:
Thanks for this suggestion.  I adjusted all the shared reshaping into helper methods (some static) and adjusted the test to use non-uniform data.  It now tests the encoding shape, that each spatial point normalizes to +/- 1 for the forward encoding, the round trip, and that we recover the correct shape and data for a non-uniform decode input.